ocaml/xapi/xapi_pool_patch.ml

```diff
@@ -36,414 +30,71 @@ open D
-  let run_path = extract_patch path in
-  finally
-    (fun () ->
-       with_logfile_fd "patch"
-         (fun log_fd ->
-            let pid = safe_close_and_exec None (Some log_fd) (Some log_fd) [] run_path args in
-            waitpid_fail_if_bad_exit pid)
-    )
-    (fun () -> Unixext.unlink_safe run_path)
-
-let guidance_from_string = function
-  | "restartHVM"  -> `restartHVM
-  | "restartPV"   -> `restartPV
-  | "restartHost" -> `restartHost
-  | "restartXAPI" -> `restartXAPI
-  | _ -> raise Bad_patch_info
-
-let precheck_patch_uuid uuid =
-  let uuid = String.lowercase uuid in
-  if not (Uuid.is_uuid uuid)
-  then raise (Invalid_patch_uuid uuid);
-  uuid
-
-let patch_info_of_xml = function
-  | Element("info", attr, _) ->
-    let find x =
-      if List.mem_assoc x attr
-      then List.assoc x attr
-      else raise (Missing_patch_key x) in
-    let label = find "name-label"
-    and descr = find "name-description"
-    and version = find "version"
-    and uuid = precheck_patch_uuid (find "uuid")
-    and guidance = find "after-apply-guidance"
-    in
-    let guidance =
-      if guidance <> "" then
-        let guidance = String.split ' ' guidance in
-        List.map guidance_from_string guidance
-      else
-        []
-    in
-    { uuid = uuid; name_label = label; name_description = descr;
-      version = version; after_apply_guidance = guidance }
-  | _ -> raise Bad_patch_info
-
-let patch_info_of_string s =
-  let xml = Xml.parse_string s in
-  debug "xml: %s" (Xml.to_string xml);
-  patch_info_of_xml xml
-
-let get_patch_info path =
-  match execute_patch path [ "info" ] with
-  | Success(output, _) ->
-    begin
-      try
-        debug "executing patch successful; parsing XML" ;
-        patch_info_of_string output
-      with e ->
-        raise (Api_errors.Server_error(Api_errors.invalid_patch, [Printexc.to_string e]))
-    end
-  | Failure(log, exn) ->
-    debug "error from patch application: %s" log;
-    begin
-      match exn with
-      | Subprocess_failed 2 ->
-        debug "probably bad line endings...";
-        raise (Api_errors.Server_error(Api_errors.invalid_patch_with_log, ["Bad line endings?"]))
-      | _ -> raise exn
-    end
-
-let read_in_and_check_patch length s path =
-  try
-    debug "Will stream patch to file: %s" path;
 
-    (* Stream the contents to path *)
-    begin
-      match length with
-      | None  ->
-        Unixext.with_file path [ Unix.O_WRONLY; Unix.O_CREAT ] 0o440
-          (fun fd -> let (_: int64) = Unixext.copy_file s fd in ())
-      | Some i ->
-        Unixext.with_file path [ Unix.O_WRONLY; Unix.O_CREAT ] 0o440
-          (fun fd -> let (_: int64) = Unixext.copy_file ~limit:i s fd in ())
-    end;
-
-    debug "Streaming complete; executing gpg";
-
-    let run_path = extract_patch path in
-    Unixext.unlink_safe run_path
+let pool_patch_of_update ~__context update_ref =
```

Comment:
An alternative implementation would be to to get the update uuid from the update_ref, then call patch_uuid_of_update_uuid, then look up the patch by its uuid.
Might be more efficient (but perhaps it doesn't matter). Arguably less elegant.